### ISSUE TITLE
Mapping <c-[>, <c-m> or <c-i> no longer works

### ISSUE THREAD

**Issue by brphilly** [neovim/neovim#17867]
### Neovim version (nvim -v)

NVIM v0.7.0-dev+1327-ga70b39965

### Vim (not Nvim) behaves the same?

no, vim 8.2.4464

### Operating system/version

Arch Linux

### Terminal name/version

kitty 0.24.4 or alacritty 0.10.1

### $TERM environment variable

xterm-kitty or alacritty

### Installation

neovim-git from AUR

### How to reproduce the issue

`nvim --clean`
Type `:nnoremap <c-[> :echo 'ctrl-sb'<cr>` and hit enter
Now press `ctrl-[` or `Esc` on your keyboard. Nothing will happen.
Now type `:nnoremap <esc> :echo 'esc'<cr>` and hit enter
Now pressing either `Esc` or `ctrl-[` on your keyboard will echo 'esc'.

### Expected behavior

I'm guessing this is caused by #17825 or one of the follow ups.

I assume my terminal emulator doesn't support distinguishing the keys (I've tried kitty and alacritty). In that case I would expect the old behaviour i.e. `:nnoremap <c-[>` works as if you had done `:nnoremap <esc>`.

### Actual behavior

Currently, mapping `<c-[>`, `<c-m>` and `<c-i>` seems to do nothing.

Mapping `<esc>`, `<cr>` and `<tab>` does work and both keys will trigger the mapping. Of course this means I could just change my mappings to use `<esc>` instead of `<c-[>` but I would rather not because in the future I might switch to a terminal emulator which can distinguish between these keys and then I would have to change the mappings back again.

**Comment by zeertzjq** [comment]
For now you can work around this by mapping `<esc>` to `<c-[>`, etc..

**Comment by zeertzjq** [comment]
Vim deals with this by defining two rounds of mappings for keys that can be simplified. This, however, only allows distinguishing these keys if the mapping for the simplified by is defined after the one for the unsimplified key.

**Comment by zeertzjq** [comment]
It is still possible to cherry-pick changes from these Vim patches:
<https://github.com/vim/vim/commit/459fd785e4a8d044147a3f83a5fca8748528aa84>
<https://github.com/vim/vim/commit/ebe9d34aa07037cff2188a8dd424ee1f59cbb0bf>
<https://github.com/vim/vim/commit/fccd93f0917234b962ce07d1df3adf9d7105936f>
<https://github.com/vim/vim/commit/e3d1f4c982bd0fe05496448d7868268c75ff7bfb>

**Comment by bfredl** [comment]
Is actually `<c-h>` broken? or does this affect the other three pairs of #17825?

**Comment by zeertzjq** [comment]
`<C-H>` is not broken.

**Comment by zeertzjq** [comment]
Many terminals send `\x08` for `<C-H>`, `\x7F` for `<BS>`, and `CSI 3 ~` for `<Del>`, which is very different from the case of `<C-I>`, `<C-M>`, and `<C-[>`.

**Comment by zeertzjq** [comment]
Also, for kitty you can try adding this to your config:

In Vim script:
```vim
if $TERM ==# 'xterm-kitty'
  autocmd UIEnter * if v:event.chan ==# 0 | call chansend(v:stderr, "\x1b[>1u") | endif
  autocmd UILeave * if v:event.chan ==# 0 | call chansend(v:stderr, "\x1b[<1u") | endif
endif
```

In Lua:
```lua
if vim.env.TERM == 'xterm-kitty' then
  vim.cmd([[autocmd UIEnter * if v:event.chan ==# 0 | call chansend(v:stderr, "\x1b[>1u") | endif]])
  vim.cmd([[autocmd UILeave * if v:event.chan ==# 0 | call chansend(v:stderr, "\x1b[<1u") | endif]])
end
```

**Comment by zeertzjq** [comment]
I took a look at these patch and it is possible to achieve the following:
- These key pairs are always treated differently if typed and in the lhs of mappings, but are treated the same in other cases (e.g. in the rhs of mappings and in Vim script string literals).

**Comment by zeertzjq** [comment]
I took a another look at these patches and found out that porting these patches cannot achieve the purpose of distinguishing these keys in lhs of mappings regardless of the order in which the mappings are defined.

**Comment by tiagovla** [comment]
BTW, for alacritty:
```yml
# ~/.config/alacritty/alacritty.yml
key_bindings:
  - { key: I, mods: Control, chars: "\x1b[105;5u" }
```

To work on tmux, I changed 2 lines on their [source](https://github.com/tiagovla/tmux/commit/f4a851c9c1aca136ee85169ee9e52dc9e5576ac3) code and built with:
```bash
./autogen.sh && ./configure && make
```

**Comment by zeertzjq** [comment]
I've been thinking about reverting #17825 and porting these Vim patches instead. However, this only allows distinguishing these key pairs in mappings if they are defined in a specific order.

**Comment by brphilly** [comment]
One potential issue I see with having to define mappings in a specific order is sometimes mappings aren't created until a plugin is lazy loaded. Also LSP mappings are often created only when an LSP client attaches to a buffer. This could make it difficult/confusing to ensure mappings are created in the correct order.

Just to be clear, my reason for creating this issue wasn't necessarily because I want these key pairs to be distinguished in kitty/alacritty. If they don't support it then I don't expect neovim to jump through hoops to make it work. My issue was that mappings of `<c-[>`, `<c-m>` or `<c-i>` no longer triggered when pressing those keys on your keyboard. With the 0.7 release coming soon I could see this causing some confusion when these mappings stop working for kitty/alacritty users.
In my opinion, it's not the end of the world if `<c-[>` and `<esc>` continue to do the same thing in certain terminal emulators. Just as long as a mapping of `<c-[>` is triggered when pressing `ctrl-[` on your keyboard.

With that being said, I'm happily using https://github.com/neovim/neovim/issues/17867#issuecomment-1079934289 so this isn't a problem for me anymore.
However there may be users who don't follow neovim development closely and won't easily find such work arounds. I will leave it up to you guys to decide what, if anything, should be done about this before the 0.7 release and beyond.

**Comment by mdelaney** [comment]
Just to add to this, my mapping for `<c-i>` no longer works. I'm guessing due to the same change? For context, I use iTerm2 (just to emphasize that this is an issue independent of terminal).

> With the 0.7 release coming soon I could see this causing some confusion

This is where I'm at now :-)

**Comment by gpanders** [comment]
@mdelaney:

>For context, I use iTerm2 (just to emphasize that this is an issue independent of terminal).

You can enable the "Report keys using CSI u" option under Profiles > [your profile] > Keys > General.

**Comment by MRAAGH** [comment]
Encountered the same issue when using urxvt (rxvt-unicode).

**Comment by justinmk** [comment]
> In that case I would expect the old behaviour i.e. `:nnoremap <c-[>` works as if you had done `:nnoremap <esc>`.

Seriously, who cares about this? This is madness. Just map the keys.

**Comment by brphilly** [comment]
> > In that case I would expect the old behaviour i.e. `:nnoremap <c-[>` works as if you had done `:nnoremap <esc>`.
> 
> Seriously, who cares about this? This is madness. Just map the keys.
> 
> You are depending on a historical accident that can be resolved by _actually mapping the keycodes you want_. Maintaining weird workaround in Nvim core forever to save a few mappings in your config, is not a good tradeoff.

Perhaps I did not word this very clearly. What I meant was even if the terminal emulator does not support distinguishing the key pairs, I want `:nnoremap <c-[>` to still work. I don't necessarily care about keeping the old behaviour exactly I just thought that would be the simplest solution for terminal emulators that don't support distinguishing the key pairs.

Let me explain the situation before and after #17825 when using `kitty` to hopefully make the situation completely clear.
## Before
Mapping `<c-[>` triggers when pressing `ctrl-[` on your keyboard. It also triggers when pressing `Esc` on your keyboard since both keys are treated the same.
Likewise mapping `<Esc>` triggers when pressing `Esc` on your keyboard and it also triggers when pressing `ctrl-[` on your keyboard.

## After
Mapping `<c-[>` does not trigger when pressing `ctrl-[` on your keyboard (or seemingly any other key).
Mapping `<Esc>` triggers when pressing `Esc` on your keyboard and also triggers when pressing `ctrl-[` on your keyboard.

# 
Basically what I want is for mapping `<c-[>` to trigger when I press `ctrl-[` on my keyboard no matter which terminal emulator I use (either one that supports distinguishing the key pairs or one that doesn't).

**Comment by clason** [comment]
So you can restore the previous behavior by mapping `<esc>` instead of `<c-[>`? I don't see how this is a major hardship.

(Just because _your_ terminal cannot distinguish these keys, you can't seriously ask every other terminal to have the same behavior, even if they can do better!)

**Comment by clason** [comment]
And let me be clear here: This change was _necessary_ to add functionality on more modern terminals, and the minor config change needed on legacy terminals was and is deemed worth it.

**Comment by miversen33** [comment]
> So you can restore the previous behavior by mapping `<esc>` instead of `<c-[>`? I don't see how this is a major hardship.
> 
> (Just because _your_ terminal cannot distinguish these keys, you can't seriously ask every other terminal to have the same behavior, even if they can do better!)

This is a very close minded approach to a problem that was _previously_ solved.
To give an example, I had `<C-i>` mapped to a particular command (in my case, `<C-w><C-h>`) and `<Tab>` mapped to a wholey separate command (`:>>`). My `<Tab>` to indent command now does not work instead operating as `<C-w><C-h>`. 

I had no issue in using these keymappings in previous versions of nvim (and vim) up until neovim 0.7.0. I understand (as pointed out by others). The intricacies of how the terminal or program handled differentiating `<C-i>` vs `<Tab>` aren't terribly important here as again, this was a solved problem before.

If the plan is to move forward with this as in, nvim is artificially limiting what keybindings can be used because "my terminal doesn't differentiate them". This may likely be true but is irrelevant as it was solved before.

Consider adding some sort of "legacy" flag/mode that can be enabled so nvim does not alienate "older" terminal users?

> (Just because your terminal cannot distinguish these keys, you can't seriously ask every other terminal to have the same behavior, even if they can do better!)
Is that not the job of the consuming application (neovim in this case) to handle?

**Comment by zeertzjq** [comment]
> To give an example, I had `<C-i>` mapped to a particular command (in my case, `<C-w><C-h>`) and `<Tab>` mapped to a wholey separate command (`:>>`). My `<Tab>` to indent command now does not work instead operating as `<C-w><C-h>`.
Note that before v0.7.0 it is likely that your `<C-i>` mapping replaced your `<Tab>` mapping.

> Consider adding some sort of "legacy" flag/mode that can be enabled so nvim does not alienate "older" terminal users?

This is easy. Actually it is my initial implementation in #17236.

**Comment by miversen33** [comment]
>  Note that before v0.7.0 it is likely that your <C-i> mapping replaced your <Tab> mapping.
I can't speak to what was happening under the hood, I can simply attest that `<C-i>` and `<Tab>` functioned differently (both in normal mode).

I appreciate the neovim team looking into how to deal with this though regardless! You're all the best

**Comment by mdelaney** [comment]
> > Just to add to this, my mapping for `<c-i>` no longer works.
> 
> @mdelaney What mapping exactly? And what key are you sending to nvim to trigger it?

@justinmk - I have this mapped to select the right window pane (I use the Colemak layout, so `i` is where `l` typically is so it gets used heavily for navigation).
```lua
vim.keymap.set("n", "<C-i>", "<C-W>l", {noremap = true, silent = true})
```

---

I get that this is fixing a long-standing problem being fixed but it is still a breaking change that's confusing from a user perspective (well, at least some users). Is there a place in the documentation that explains this (if not, I might be able to contribute once I understand this a bit better)? It just seems like this could be a common point of confusion. It'd be really useful if there was a "common mapping issues" page somewhere (again, I'd be willing to spend a bit of time on this if you think it'd be useful here as well).

(I didn't see anything obvious explaining this in [the map doc page](https://neovim.io/doc/user/map.html))

**Comment by justinmk** [comment]
> I have this mapped to select the right window pane (I use the Colemak layout, so `i` is where `l` typically is so it gets used heavily for navigation).
> 
> ```lua
> vim.keymap.set("n", "<C-i>", "<C-W>l", {noremap = true, silent = true})
> ```

If typing "CTRL" and "i" doesn't trigger that mapping, then that is a Nvim bug. If you are expecting `<tab>` to also trigger it, then no, that's not a bug.

**Comment by zeertzjq** [comment]
Make some small divergence when porting Vim patch 8.1.2145 may make this work. I may continues #17932 later.

**Comment by zeertzjq** [comment]
> Making some small divergences when porting Vim patch 8.1.2145 may make this work. I may continue #17932 later.

I though again and I cannot make this divergence. It will make `feedkeys()` behave very confusingly.

**Comment by zeertzjq** [comment]
So again there are only two options here:
1. Fully match Vim's behavior and only distinguish `<C-I>` and `<Tab>` in mappings if a `<Tab>` mapping is defined after a `<C-I>` mapping
2. Keep the current behavior, which completely diverges from Vim's behavior and means this issue is wontfix

**Comment by zeertzjq** [comment]
This is probably to complicated for v0.7.1.

**Comment by zeertzjq** [comment]
The PR is probably too large to backport to v0.7.1.

**Comment by dev-uwv** [comment]
> BTW, for alacritty:
> 
> # ~/.config/alacritty/alacritty.yml
> key_bindings:
>   - { key: I, mods: Control, chars: "\x1b[105;5u" }
> To work on tmux, I changed 2 lines on their [source](https://github.com/tiagovla/tmux/commit/f4a851c9c1aca136ee85169ee9e52dc9e5576ac3) code and built with:
> 
> ./autogen.sh && ./configure && make

With your help I made de CTRL+I work in lazyvim

```
# File: alacritty.toml 
# For alacritty TAB is the same as "Ctrl + i". Add `nnoremap <C-I> <C-I>` to neovim.
{ key = "I", mods = "Control", chars = "\u001B[105;5u" },

```
```
-- File: keymaps.lua
-- Navigate buffers
vim.keymap.set("n", "<Tab>", ":bnext<CR>", { silent = true, noremap = true })
vim.keymap.set("n", "<S-Tab>", ":bprevious<CR>", { silent = true, noremap = true })

-- For alacritty TAB is the same as "Ctrl + i". Therefor use `nnoremap <C-I> <C-I>`
vim.keymap.set("n", "<C-i>", "<C-i>", { silent = true, noremap = true })

```
